Batchnorm now always updates var and mean inplace

VitalyFedyunin opened the pull request:
Fixes: #33867

Also fixes similar case for miopen.

Do nothing with version counters.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#38040 Batchnorm now always updates var and mean inplace**

Differential Revision: [D21460273](https://our.internmc.facebook.com/intern/diff/D21460273)

VitalyFedyunin committed:
Batchnorm now always updates var and mean inplace

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "Batchnorm now always updates var and mean inplace"

Fixes: #33867

Also fixes similar case for miopen.

Do nothing with version counters.


Differential Revision: [D21460273](https://our.internmc.facebook.com/intern/diff/D21460273)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/Normalization.cpp:
```diff
@@ -491,10 +491,27 @@ Tensor batch_norm(
+      momentum,
+      eps,
+      cudnn_enabled);
+  // Not all implementations of batch norm do inplace modifications of mean and var #33867
```
This seems like a pessimization? As in, even if there is an inplace modification, you're always doing an extra `copy_` even if it isn't necessary. I feel like it would be better to fix this at the backend level, whenever appropriate.

VitalyFedyunin replied:
Nope, if backend does proper in-place `std::get<1>(result)` tensor would be the same as `running_mean` tensor, and copy_ will fast exit.

ngimel commented on aten/src/ATen/native/Normalization.cpp:
```diff
@@ -491,10 +491,27 @@ Tensor batch_norm(
+      eps,
+      cudnn_enabled);
+  // Not all implementations of batch norm do inplace modifications of mean and var #33867
+  if (running_mean.defined()) {
```
Returned tuple has a following structure:
```
return std::tuple<Tensor, Tensor, Tensor, Tensor>{output_t, save_mean, save_var, reserve};
```
save_mean and save_var are not running_mean/running_var, so when you are getting 1st and 2nd elements of the returned tuple, you don't get what you think you are getting,

VitalyFedyunin replied:
Changed

VitalyFedyunin committed:
Update on "Batchnorm now always updates var and mean inplace"

Fixes: #33867

Also fixes similar case for miopen.

Do nothing with version counters.


Differential Revision: [D21460273](https://our.internmc.facebook.com/intern/diff/D21460273)

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "Batchnorm now always updates var and mean inplace"

Fixes: #33867

Also fixes similar case for miopen.

Do nothing with version counters.


Differential Revision: [D21460273](https://our.internmc.facebook.com/intern/diff/D21460273)

[ghstack-poisoned]

VitalyFedyunin commented:
`test_batchnorm_simple_average_cuda_float32` test failures looks related, I will investigate.
